#!/bin/bash -e
pkg=memtest86+-4.10
tar xzf ${1}/${pkg}.tar.gz
cd ${pkg}
patch -p1 <<"EOF"
--- memtest86+/Makefile	2010-04-28 07:11:15.000000000 -0500
+++ memtest86+/Makefile	2010-06-28 16:48:48.352193491 -0500
@@ -11,7 +11,7 @@
 AS=as -32
 CC=gcc
 
-CFLAGS= -Wall -march=i486 -m32 -O2 -fomit-frame-pointer -fno-builtin -ffreestanding -fPIC -fno-stack-protector
+CFLAGS= -Wall -march=i486 -m32 -O2 -fomit-frame-pointer -fno-builtin -ffreestanding -fPIC
 
 OBJS= head.o reloc.o main.o test.o init.o lib.o patn.o screen_buffer.o \
       config.o linuxbios.o memsize.o pci.o controller.o random.o spd.o \
diff -uNPr memtest86+/README.Xiotech memtest86+/README.Xiotech
--- memtest86+/README.Xiotech	1969-12-31 18:00:00.000000000 -0600
+++ memtest86+/README.Xiotech	2008-02-19 17:58:25.000000000 -0600
@@ -0,0 +1,2 @@
+This came from memtest86+-4.10. It has been modified only to support the
+serial interface.
diff -uNPr memtest86+/config.h memtest86+/config.h
--- memtest86+/config.h	2008-02-19 17:57:50.000000000 -0600
+++ memtest86+/config.h	2008-02-19 17:59:36.000000000 -0600
@@ -13,16 +13,16 @@
 /* SERIAL_CONSOLE_DEFAULT -  The default state of the serial console. */
 /*	This is normally off since it slows down testing.  Change to a 1 */
 /*	to enable. */
-#define SERIAL_CONSOLE_DEFAULT 0
+#define SERIAL_CONSOLE_DEFAULT 1
 
 /* SERIAL_TTY - The default serial port to use. 0=ttyS0, 1=ttyS1 */ 
 #define SERIAL_TTY 0
 
 /* SERIAL_BAUD_RATE - Baud rate for the serial console */
-#define SERIAL_BAUD_RATE 9600
+#define SERIAL_BAUD_RATE 115200
 
 /* BEEP_MODE - Beep on error. Default off, Change to 1 to enable */
-#define BEEP_MODE 0
+#define BEEP_MODE 1
 
 /* BEEP_END_NO_ERROR - Beep at end of each pass without error. Default off, Change to 1 to enable */
 #define BEEP_END_NO_ERROR 0
diff -uNPr memtest86+/memtest.lds memtest86+/memtest.lds
--- memtest86+/memtest.lds	2008-02-19 17:57:50.000000000 -0600
+++ memtest86+/memtest.lds	2008-02-19 18:00:13.000000000 -0600
@@ -3,7 +3,7 @@
 
 ENTRY(_start); 
 SECTIONS {
-	. = 0x5000;
+	. = 0x100000;
 	_start = . ;
 	.data : {
 		*(.data)
diff -uNPr memtest86+/setup.S memtest86+/setup.S
--- memtest86+/setup.S	2008-02-28 17:07:32.000000000 -0600
+++ memtest86+/setup.S	2008-02-28 15:50:26.000000000 -0600
@@ -42,7 +42,7 @@
 	jz	alt_a20_done
 	
 	/* set or clear bit1, the ALT_A20_GATE bit */
-	movb	4(%esp), %ah
+	addr32 movb	4(%esp), %ah
 	testb	%ah, %ah
 	jz	alt_a20_cont1
 	orb	$2, %al
EOF
echo "Patch of ${pkg} completed"
# End of patch
